### PR TITLE
Refactor NRL h2o photochemical physics scheme

### PR DESCRIPTION
**Pull request by dustinswales**
Included in this PR are changes to NRL's stratospheric h2o photochemistry scheme to make it "safe across multiple CCPP instances". These changes are identical to the reorganization that [o3 photochemistry scheme underwent in the fall of 2023.](https://github.com/ufs-community/ccpp-physics/pull/75). 

Both schemes are called from a new module, GFS_photochemistry, at the start of the time-split section of the suite. Previously, the ozone scheme was called from GFS_stateout_update, and the h2o scheme was called just after.

Question @grantfirl @climbfuji 
The o3 and h2o photochemistry schemes are both "time-splitting". Maybe this is a good place to start with the cleanup for MPAS effort (e.g. enforcing the procces-splitting requirement)?

### CODE REVIEW

**Review by grantfirl — APPROVED** [review]
Nice job, looks good. Does this need to be high priority for some reason?

**Comment by dustinswales** [comment]
No rush on this.

**Comment by climbfuji** [comment]
> Question @grantfirl @climbfuji The o3 and h2o photochemistry schemes are both "time-splitting". Maybe this is a good place to start with the cleanup for MPAS effort (e.g. enforcing the proccess-splitting requirement)?

I would not try to attempt commingling the changes from @michakales and framework-related development.

With this PR and the previous PRs for ccpp-framework and ccpp-physics, are all changes from the original PR included? Or is there still something missing?

File: physics/Interstitials/UFS_SCM_NEPTUNE/GFS_photochemistry.F90 [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+         dp,           & ! Pressure thickness (Pa)
+         gt0             ! Air temperature (K)
+    real(kind=kind_phys), intent(in), dimension(:,:,:) :: &
+         ozpl,         & ! Ozone data for current model timestep.
+         h2opl           ! h2o data for curent model timestep.
+    logical,              intent(in) :: &
+         oz_phys_2015, & ! Do ozone photochemistry? (2015)
+         oz_phys_2006, & ! Do ozone photochemistry? (2006)
+         h2o_phys        ! Do stratospheric h2o photochemistry?
+    type(ty_ozphys), intent(in) :: &
```
Comment (climbfuji) [inline physics/Interstitials/UFS_SCM_NEPTUNE/GFS_photochemistry.F90:71]: Are ozphys and/or h2ophys conditionally allocated, i.e. do they need to be optional here?

File: physics/Interstitials/UFS_SCM_NEPTUNE/GFS_photochemistry.F90 [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+         errflg          ! CCPP Error flag.
+
+    ! Initialize CCPP error handling variables
+    errmsg = ''
+    errflg = 0
+
+    ! If no photchemical scheme is on, but SDF has this module, report an error?
+    ! DJS2024 Asks: If everything in _run() is controlled by these logicals, we probably don't
+    !         need to kill it here, since the run phase will be harmless?
+    if ((.not. oz_phys_2006) .and. (.not. oz_phys_2015) .and. (.not. h2o_phys)) then
```
Comment (climbfuji) [inline physics/Interstitials/UFS_SCM_NEPTUNE/GFS_photochemistry.F90:32]: The logic in `_run` in principle allows you to run both oz_phys_2006 and oz_phys_2015. Maybe a good idea would be to check in the `_init` routine that not both are enabled.

File: physics/Interstitials/UFS_SCM_NEPTUNE/GFS_photochemistry.meta [hunk 3]
```diff
@@ -0,0 +1,210 @@
...
+  type = character
+  kind = len=*
+  intent = out
+[errflg]
+  standard_name = ccpp_error_code
+  long_name = error code for error handling in CCPP
+  units = 1
+  dimensions = ()
+  type = integer
+  intent = out
```
Comment (climbfuji) [inline physics/Interstitials/UFS_SCM_NEPTUNE/GFS_photochemistry.meta]:
```suggestion
  intent = out

```

File: physics/Interstitials/UFS_SCM_NEPTUNE/GFS_suite_stateout_update.meta [hunk 4]
```diff
@@ -2,8 +2,8 @@
 [ccpp-table-properties]
   name = GFS_suite_stateout_update
   type = scheme
-  dependencies = ../../hooks/machine.F,../../photochem/module_ozphys.F90
-
+  dependencies = ../../hooks/machine.F
+
```
Comment (climbfuji) [inline physics/Interstitials/UFS_SCM_NEPTUNE/GFS_suite_stateout_update.meta]:
```suggestion

```

File: physics/photochem/module_h2ophys.meta [hunk 5]
```diff
@@ -0,0 +1,24 @@
...
+
+[ccpp-arg-table]
+  name = module_h2ophys
+  type = module
+[ty_h2ophys]
+  standard_name = ty_h2ophys
+  long_name = definition of type ty_h2ophys
+  units = DDT
+  dimensions = ()
+  type = ty_h2ophys
```
Comment (climbfuji) [inline physics/photochem/module_h2ophys.meta]:
```suggestion
  type = ty_h2ophys

```

**Review by climbfuji — CHANGES_REQUESTED** [review]

**Comment by dustinswales** [comment]
> > Question @grantfirl @climbfuji The o3 and h2o photochemistry schemes are both "time-splitting". Maybe this is a good place to start with the cleanup for MPAS effort (e.g. enforcing the proccess-splitting requirement)?
> 
> I would not try to attempt commingling the changes from @michakales and framework-related development.
> 
> With this PR and the previous PRs for ccpp-framework and ccpp-physics, are all changes from the original PR included? Or is there still something missing?

@climbfuji The only piece missing from the OG PR is to move the Thompson mp "is_initialized" flag from module memory to GFS_typedefs. (This will avoid needing to bring in the "instance" field into the physics schemes.

File: physics/Interstitials/UFS_SCM_NEPTUNE/GFS_photochemistry.F90 (hunk 1 above)
Comment (dustinswales) [reply physics/Interstitials/UFS_SCM_NEPTUNE/GFS_photochemistry.F90:71]: They are required by the scheme and are not optional from the physics POV.

File: physics/Interstitials/UFS_SCM_NEPTUNE/GFS_photochemistry.F90 (hunk 2 above)
Comment (dustinswales) [reply physics/Interstitials/UFS_SCM_NEPTUNE/GFS_photochemistry.F90:32]: Good call. Will do.

**Review by Qingfu-Liu — APPROVED** [review]

**Comment by grantfirl** [comment]
@dustinswales Looks like there are some comments that need to be addressed before this is put in the queue?

**Comment by dustinswales** [comment]
@grantfirl Yeah. I've been meaning to circle back to this. I will update and address the comments later this afternoon.

**Comment by grantfirl** [comment]
@climbfuji Please re-review and see if https://github.com/ufs-community/ccpp-physics/pull/213/commits/420317e6b8e4adb57c68b1ea9381d1a15f4cc86f addresses your concerns.

**Review by RuiyuSun — APPROVED** [review]

**Review by climbfuji — APPROVED** [review]
Looks good to me, thanks very much!

**Comment by grantfirl** [comment]
@dustinswales I'm thinking of combining this with https://github.com/ufs-community/ccpp-physics/pull/218 to expedite testing/merging since this doesn't chance baselines. Is that OK with you?

File: physics/MP/Thompson/mp_thompson.meta [hunk 6]
```diff
@@ -282,6 +282,12 @@
   kind = kind_phys
   intent = in
   optional = True
+[is_initialized]
+  standard_name = flag_for_thompson_mp_scheme_initialization
+  long_name = flag carrying scheme initialization status
+  units = flag
```
Comment (grantfirl) [inline physics/MP/Thompson/mp_thompson.meta:288]: Same thing here. Missing dimensions causing ccpp_prebuild to fail.

Comment (dustinswales) [reply physics/MP/Thompson/mp_thompson.meta:288]: Sorry, I had un committed changes. My bad.

**Comment by grantfirl** [comment]
@dustinswales Combined into https://github.com/ufs-community/ccpp-physics/pull/223 (Scheduled for UFS merge queue early next week)

**Pull request closed without merging**